Changes to IFuelHandler making it possible to edit vanilla burn times

jaredlll08 opened the pull request:
Using this, mods can dictate whether their IFuelHandler handles an item or not, as well as set the burn time of vanilla items to 0.

mezz committed:
Fix Universal bucket handling for Fluids with NBT (Fix #3757)

mezz committed:
Update mcp mappings to 20170611

mezz committed:
Clean up ShapedOreRecipe and ShapelessOreRecipe

jaredlll08 committed:
Added a new method to IFuelHandler that will check if an item is handled

liach commented:
Make the original method return an OptionalInt instead.

Alpvax commented on src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java:
```diff
@@ -341,11 +341,21 @@ public static int getFuelValue(@Nonnull ItemStack itemStack)
+        boolean handled = false;
+        for(IFuelHandler handler : fuelHandlers) {
+            if(handler.matches(itemStack))
+                handled = true;
```
You should just return true here, and return false after the for loop. If it has already matched, you can stop searching and avoid the overhead.

diesieben07 commented:
You can already do this if you return -1 from your handler. -1 will always be the smaller one of the two values in the `max` call, and thus be ignored.

jaredlll08 commented:
@diesieben07 while that is true, it still doesn't help with changing vanilla burn times

jaredlll08 committed:
Changed IFuelHandler to use OptionalInt

liach commented on src/main/java/net/minecraftforge/fml/common/registry/GameRegistry.java:
```diff
@@ -341,19 +341,18 @@ public static int getFuelValue(@Nonnull ItemStack itemStack)
         for(IFuelHandler handler : fuelHandlers) {
-            if(handler.matches(itemStack))
-                handled = true;
+            if(handler.getBurnTime(itemStack) !=null && handler.getBurnTime(itemStack).isPresent())
```
That method is not declared nonnull so I don't suggest you to have the null check.

jaredlll08 committed:
Removed null check

liach commented:
I don't know if some addon mods want to override fuel values in some other mods. I attempted to do that in #2691 but then failed.

jaredlll08 commented:
Yea, that is actually why I'm making this PR, Crafttweaker allows users to change fuel values, and it is a bit inconvenient not being able to change vanilla values

liach commented:
I wonder if we should introduce a priority system into the fuel handling.

Darkhax commented:
At this point, I think switching this over to an event system may be the way to go. Similar to how we use events for the anvil rather than an actual anvil recipe system.

cpw committed:
Patch a small bug in SPacketLoginSuccess that lets us emulate the network
handshake. This will be very useful as I rebuild the networking.

shadowfacts committed:
Allow condition factories to be registered from JSON (#4015)

mezz commented on patches/minecraft/net/minecraft/tileentity/TileEntityFurnace.java.patch:
```diff
@@ -62,19 +62,18 @@
          {
-             Item item = p_145952_0_.func_77973_b();
-+            if (!item.getRegistryName().func_110624_b().equals("minecraft"))
++            if(net.minecraftforge.fml.common.registry.GameRegistry.isFuelHandled(p_145952_0_))
```
I think I'd prefer you did this without calling everything twice. 
Just get an `OptionalInt` from the `GameRegistry` once and if it exists use it here.
No need for `isFuelHandled`.

mezz committed:
Update MCP mappings to 20170617

mezz committed:
Add missing Override annotations

jaredlll08 committed:
Moved back to returning an int, using -1 as ignored

mezz commented on src/test/java/net/minecraftforge/debug/FuelHandlerTest.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Debug mod that sets the burn time of every item to 0
+ */
+@Mod(modid = "fuelhandlertest", name = "Guel Handler Test", version = "0.0.0")
```
"Guel"

mezz replied:
Use acceptable remote versions = `*` so that this can run when connected to vanilla clients or servers.

mezz commented on patches/minecraft/net/minecraft/tileentity/TileEntityFurnace.java.patch:
```diff
@@ -62,19 +62,21 @@
-             Item item = p_145952_0_.func_77973_b();
-+            if (!item.getRegistryName().func_110624_b().equals("minecraft"))
++            int burnTime = net.minecraftforge.fml.common.registry.GameRegistry.getFuelValue(p_145952_0_);
++            if(burnTime >= 0)
```
should have a space between `if` and the parenthesis.

mezz commented on patches/minecraft/net/minecraft/tileentity/TileEntityFurnace.java.patch:
```diff
@@ -62,19 +62,21 @@
 +            }
- 
+             Item item = p_145952_0_.func_77973_b();
+-
```
The "clean" version has a newline here after the `Item item =` line, you should not remove it.

mezz replied:
@jaredlll08 I think this patch still needs fixing

mezz commented on src/test/java/net/minecraftforge/debug/FuelHandlerTest.java:
```diff
@@ -0,0 +1,21 @@
+import net.minecraftforge.fml.common.event.FMLInitializationEvent;
+import net.minecraftforge.fml.common.registry.GameRegistry;
+
+import java.util.OptionalInt;
```
Remove unused imports

mezz commented on src/test/java/net/minecraftforge/debug/FuelHandlerTest.java:
```diff
@@ -0,0 +1,21 @@
+package net.minecraftforge.debug;
+
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.fml.common.*;
```
Do not import with `*`

liach commented on src/main/java/net/minecraftforge/fml/common/IFuelHandler.java:
```diff
@@ -23,5 +23,12 @@
+     * @param fuel ItemStack to get the burn time of
+     * @return the burn time of the ItemStack, or -1 for ignored
+     */
     int getBurnTime(ItemStack fuel);
```
Why do we keep it? We should depreate this and use `ToIntFunction<ItemStack>`

jaredlll08 replied:
I am simply doing what @mezz requested.

jaredlll08 committed:
Fixed requests changes

HenryLoenwind commented:
Has nobody noticed that not being able to change vanilla data here is a feature, not a bug?

jaredlll08 commented:
Please enlighten me as to how not being able to change vanilla burn times is a feature?

Alpvax commented:
Discussion as part of #3627

mezz committed:
Add timeout to TestNetworkHandshake setup

mezz committed:
Disable TestNetworkHandshake to make Travis work again

mezz committed:
Log info on Travis tests

mezz committed:
Fix itemstack comparison with null

LexManos committed:
First pass of registry rewrite.
Tests do not run yet.
Things of note:
Removed the idea of substitutions, just register multiple items with the same name they will override each other like a normal map.
Decoupled Forge registries from vanilla classes. They now use bouncer classes. MODDERS SHOULD NEVER USE THESE
Introduced more stringent registry locking. As it sits things are only allowed to be registered during the RegistryEvent.Registry phases!
This is to force modders to split up their registrations, and pave the way for calling these functions multiple times during a single MC lifecycle.
ObjectHolder, Missing Mappings, etc... *should* now work for all registry types, nothing should be special cased to Blocks/Items anymore.
Added optional generic dummy factory to registries, to allow registry creators the ability to dummy more then just Blocks.
Re-worked the ItemStack transformer and applied it to ItemBlock and ItemBlockSpecial. Allowing us to cleanup reflective hacks in Block callbacks.
Registry onAdd callback is now only fired on the ACTIVE registry, fixing any issues of temp registries overriding active objects.

bs2609 committed:
Remove unused patch to ItemRecord. Closes #2686. (#4035)

lumien231 committed:
[1.12] Fixes Exception when getting modlist for a vanilla client (#4028)

williewillus committed:
Fix shapeless recipes not being able to use NBT on output, and fixed issue with null groups.

LexManos committed:
Fix loading recipe jsons from jar files.

mezz committed:
Fix patches in Chunk and Explosion (#4036)

TechnicianLP committed:
IRecipeConditions forge:and and forge:false (#4017)

bs2609 committed:
Remove Java 8 warnings now that Minecraft requires it

LexManos committed:
More work, game loads and runs now.
Moved Registry events to directly AFTER PreInit instead of before.
This allows modders to register handlers for it without @EventBusSubscriber.
It also allows you to register custom things needed before the construction of blocks or items such as Fluids.
TODO: Move Fluids to a real registry.
@ObjectHolder can now be used on private fields.
Reworked FMLModIdMappingEvent to include data for all registries.
Tile Entities are now registrable.

mezz committed:
Fix Chunk  patch (#4036)

mezz commented:
Right, unless Lex changed his mind you will need to bypass the Forge handler for items from the `minecraft` domain.

LexManos committed:
Fixup some registry tests, dummies are now persisted across syncing.

bs2609 committed:
Update NoteBlockEvent with new instruments. Resolves #4018. (#4019)

LexManos committed:
Fix recipe names in packaged jars. Closes #4039

jaredlll08 commented:
Mods such as CraftTweaker allow modpacks authors, server owners and players to change the fuel values.
If a modpack author wants to change it that items in their pack only burn for half of their usual time, making players grind more resources, they can easily do so, however, as it currently stands, and the reason this PR was actually created, it is impossible to change vanilla burn time, which could completely break the progression in a modpack.

mezz committed:
Fix ShapedOreRecipe matching. #4038

LexManos committed:
Disable substitution test, substitutions are not managed the way they used to be.
We do not care about substitutions, People who replace entries should make them compatible with the existing item.

diesieben07 committed:
Clean up Logging (#3953)

# Conflicts:
#	patches/minecraft/net/minecraft/client/renderer/texture/Stitcher.java.patch
#	patches/minecraft/net/minecraft/client/renderer/texture/TextureMap.java.patch
#	patches/minecraft/net/minecraft/entity/EntityList.java.patch
#	patches/minecraft/net/minecraft/tileentity/TileEntity.java.patch
#	src/main/java/net/minecraftforge/oredict/OreDictionary.java

LexManos committed:
Implement override tracking and syncing.
Overrides MUST be registered within a tracked event so we know what mod it came from.
This will allow servers/saves to select which mod 'wins' and becomes to active entry.
This should also mean that when connecting to a vanilla server things will revert.

LexManos committed:
Fix conflicts

LexManos committed:
Loosen up registry locks until data freeze.

tterrag1098 committed:
Make ForgeRegistry iterator sorted. Closes #4058 (#4059)

bs2609 committed:
Fire AnimalTameEvent for parrots (#4003)

mezz committed:
Update MCP to 20170624

TechnicianLP committed:
Fix onBlockClicked handler firing twice and Item.onArmorTick in the wrong place (#4037)

Choonster committed:
Fix uncaught ClassCastException from invalid recipe file (#4043)

A JsonSyntaxException will now be thrown, caught and logged when the conditions property of a recipe isn't an array

bs2609 committed:
[1.12] Cleanup of files with old licence headers, Add missing LGPL 2.1 licence headers (#4051)

KingLemming committed:
Use instanceof ItemFishingRod instead of item equality (#4060)

mezz committed:
Clean up logging

mezz committed:
Remove arrival/departure message methods from WorldProvider (#2733)

mezz committed:
Add Block.getDrops that uses NonNullList, deprecate the old method. Fix #3567 (#3778)

LexManos committed:
Fix minecraft:item_nbt ingredient type.

LexManos committed:
Fix Emerald generation causing chunk loading.

LexManos committed:
Make getSubItems non-client only. Use it for Ingredient iteration.

LexManos committed:
Fix NPE in MissingMappings event. Closes #4067

LexManos committed:
Make Dummy block an Air block.

LexManos committed:
Reload client recipe book after loading modded recipes.

RainWarrior committed:
Cleanup of model system, taking advantage of java 8;
replaced guava Optional and Function with java 8 versions.

LexManos committed:
Make Block.getSubBlocks not client side only.

LexManos committed:
Fix ModelBlockAnimation expsing a private class.

LexManos committed:
Fix missed side only on BlockFlower

jaredlll08 committed:
Removed line

jaredlll08 committed:
Merge remote-tracking branch 'origin/1.12.x' into 1.12.x

mezz commented:
After discussing with jared and with lex, I wrote a new version of this PR here: https://github.com/MinecraftForge/MinecraftForge/pull/4083
